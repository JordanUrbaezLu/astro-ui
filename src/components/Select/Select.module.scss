@use "../../utility/globals";

.selectContainer {
  font-family: globals.$font-family;
  position: relative;
  user-select: none;
  width: 142px;
}

.select {
  align-items: center;
  border: 3px solid globals.$primary-background-color;
  border-radius: 5px;
  column-gap: 6px;
  display: flex;
  font-size: 16px;
  outline: none !important;
  padding: 10px 6px;
}

.selectInput {
  align-items: center;
  border: 2px solid globals.$primary-background-color;
  border-radius: 6px;
  box-sizing: border-box;
  display: flex;
  font-size: 18px;
  height: 46px;
  padding: 10px;
  transition: all 130ms ease-in;
  width: 142px;

  &.isOpen {
    box-shadow: 0 0 5px 5px globals.$primary-box-shadow;
  }
}

.label {
  color: globals.$primary-background-color;
  font-family: globals.$font-family;
  font-size: 14px;
  padding-left: 8px;
  user-select: none;
}

.selectIcon {
  background-color: transparent !important;
  color: globals.$primary-background-color !important;
  left: 100px;
  position: absolute;
  top: 4px;
  z-index: 1;

  &:hover,
  &:focus {
    background-color: globals.$primary-text-color !important;
  }

  &:active {
    background-color: globals.$primary-background-color-active !important;
    color: globals.$primary-text-color !important;
  }
}

.enter {
  opacity: 0;
  transform: translateY(-6px);
}

.enterActive {
  opacity: 1;
  transform: translateY(0);
  transition: all 200ms ease-in-out;
}

.exit {
  opacity: 1;
  transform: translateY(0);
}

.exitActive {
  opacity: 0;
  transform: translateY(-6px);
  transition: all 200ms ease-in-out;
}
