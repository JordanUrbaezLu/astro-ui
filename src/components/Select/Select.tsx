import { CSSTransition } from "react-transition-group";
import styles from "./Select.module.scss";
import itemStyles from "./SelectItem.module.scss";
import { IoChevronDownSharp, IoChevronUpSharp } from "react-icons/io5";
import * as React from "react";
import { TextField } from "../TextField";

export interface SelectProps {
  /**
   * The content for the Select
   */
  children: React.ReactNode;
  /**
   * If the Select initially renders as open
   *
   * @default false
   */
  startsOpen?: boolean;
  /**
   * The label for the Select
   */
  label?: string;
}

export const Select: React.FC<SelectProps> = (props) => {
  const { children, label, startsOpen = false } = props;

  const [isOpen, setIsOpen] = React.useState<boolean>(startsOpen);
  const [value, setValue] = React.useState<string>("");

  return (
    <div className={styles.selectContainer}>
      <TextField
        size="large"
        selectable={false}
        onChange={() => {}}
        value={value}
        label={label}
        className={styles.inputContainer}
        onClick={() => setIsOpen(!isOpen)}
      />
      {isOpen ? (
        <IoChevronUpSharp className={styles.icon} />
      ) : (
        <IoChevronDownSharp className={styles.icon} />
      )}
      <CSSTransition
        in={isOpen}
        timeout={200}
        mountOnEnter
        unmountOnExit
        classNames={{
          enter: styles.enter,
          enterActive: styles.enterActive,
          exit: styles.exit,
          exitActive: styles.exitActive,
        }}
      >
        <div className={styles.selectOptionsMenu} role="menu">
          {React.Children.map(children, (child, index) => {
            if (
              React.isValidElement<React.HTMLAttributes<HTMLElement>>(child)
            ) {
              return React.cloneElement(child, {
                className: itemStyles.selectItem,
                key: index,
                onClick: () => {
                  setValue(child.props.children as string);
                  setIsOpen(!isOpen);
                },
              });
            } else {
              return null;
            }
          })}
        </div>
      </CSSTransition>
    </div>
  );
};
