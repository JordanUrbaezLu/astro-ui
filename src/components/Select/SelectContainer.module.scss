@use "../../utility/globals";

.selectOptionsMenu {
  background-color: globals.$primary-background-color;
  border-radius: 4px;
  display: flex;
  flex-direction: column;
  margin-top: 2px;
  padding: 4px 0;
  position: absolute;
  width: 142px;
  z-index: 1;
}
