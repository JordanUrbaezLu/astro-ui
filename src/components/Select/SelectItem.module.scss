@use "../../utility/globals";

.selectItem {
  color: white;
  cursor: pointer;
  font-family: globals.$font-family;
  font-size: 16px;
  outline: none;
  padding: 5px 5px 5px 12px;
  text-align: left;

  &.selected {
    background-color: rgb(97 70 109);

    &:hover,
    &:focus {
      background-color: rgb(110 97 115);
    }

    &:active {
      background-color: globals.$primary-background-color-active;
    }
  }

  &:hover,
  &:focus {
    background-color: globals.$primary-background-color-hover-and-focus;
  }

  &:active {
    background-color: globals.$primary-background-color-active;
  }
}
