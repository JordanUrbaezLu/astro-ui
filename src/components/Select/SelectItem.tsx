import * as React from "react";
interface SelectItemProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The content for the SelectItem
   */
  children: string;
}

/**
 * @public
 */
export const SelectItem: React.FC<SelectItemProps> = (props) => {
  const { className, children, ...rest } = props;
  return (
    <div className={className} {...rest}>
      {children}
    </div>
  );
};

SelectItem.displayName = "SelectItem";
