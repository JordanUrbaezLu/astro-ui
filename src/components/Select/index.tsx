/* istanbul ignore file */
export * from "./Select";
export * from "./SelectItem";
