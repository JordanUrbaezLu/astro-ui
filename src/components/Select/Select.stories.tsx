import { Story } from "@storybook/react";
import * as React from "react";
import { Select, SelectProps } from "./Select";
import { SelectItem } from "./SelectItem";
import { within, userEvent } from "@storybook/testing-library";
import { allowLayoutCalculations } from "../../../.storybook/utilities/allowLayoutCalculations";

export default {
  component: Select,
  parameters: {
    chromatic: { delay: 2000 },
  },
  title: "Components/Inputs/Select",
};

const Template: Story<SelectProps> = (args) => {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  return (
    <Select
      {...args}
      isOpen={isOpen}
      onOpen={() => setIsOpen(true)}
      onClose={() => setIsOpen(false)}
    >
      <SelectItem>CA</SelectItem>
      <SelectItem>MA</SelectItem>
      <SelectItem>FL</SelectItem>
    </Select>
  );
};

export const Default = Template.bind({});

export const Placeholder = Template.bind({});
Placeholder.args = {
  placeholder: "State",
};
Placeholder.play = ({ canvasElement }) => {
  const canvas = within(canvasElement);
  allowLayoutCalculations(() =>
    userEvent.click(canvas.getByRole("button"))
  );
};

export const Selected = Template.bind({});
Selected.args = {
  selected: 1,
};
Selected.play = ({ canvasElement }) => {
  const canvas = within(canvasElement);
  allowLayoutCalculations(() =>
    userEvent.click(canvas.getByRole("button"))
  );
};

export const WithManySelectItems: Story<SelectProps> = (args) => {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  return (
    <Select
      {...args}
      isOpen={isOpen}
      onOpen={() => setIsOpen(true)}
      onClose={() => setIsOpen(false)}
    >
      <SelectItem>10</SelectItem>
      <SelectItem>20</SelectItem>
      <SelectItem>30</SelectItem>
      <SelectItem>40</SelectItem>
      <SelectItem>50</SelectItem>
      <SelectItem>60</SelectItem>
      <SelectItem>70</SelectItem>
      <SelectItem>80</SelectItem>
      <SelectItem>90</SelectItem>
    </Select>
  );
};

WithManySelectItems.play = ({ canvasElement }) => {
  const canvas = within(canvasElement);
  allowLayoutCalculations(() =>
    userEvent.click(canvas.getByRole("button"))
  );
};
