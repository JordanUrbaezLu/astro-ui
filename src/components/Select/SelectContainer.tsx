import classNames from "classnames";
import * as React from "react";
import { SelectProps } from "./Select";
import styles from "./SelectContainer.module.scss";
import itemStyles from "./SelectItem.module.scss";
import { useOnClickOutside, useOnKeyDown } from "../../hooks";
import { SelectItem } from "./SelectItem";
import { useSelectFocus } from "../../hooks/useSelectFocus";

export interface SelectContainerProps extends SelectProps {
  /**
   * The callback fired when requesting to change selected index
   */
  onSetSelectedIndex: (index: number) => void;
  /**
   * The ref for the Select
   */
  selectRef: React.RefObject<HTMLDivElement>;
  /**
   * The trigger ref for the Select
   */
  triggerRef: React.RefObject<HTMLButtonElement>;
}

/**
 * @private
 */
export const SelectContainer: React.FC<SelectContainerProps> = (
  props
) => {
  const {
    children,
    selected,
    onClose,
    onSetSelectedIndex,
    selectRef,
    triggerRef,
  } = props;

  const selectContainerRef = React.useRef<HTMLDivElement>(null);

  const { focusNextItem, focusPreviousItem, focusedSelectItemIndex } =
    useSelectFocus(children, selectContainerRef, selected);

  useOnClickOutside(selectRef, onClose);
  useOnKeyDown(["Escape"], () => {
    onClose();
    triggerRef.current?.focus();
  });

  return (
    <div
      className={styles.selectOptionsMenu}
      ref={selectContainerRef}
      role="listbox"
    >
      {React.Children.map(children, (child, index) => {
        const selectItem = child as React.ReactElement;

        if (selectItem.type === SelectItem) {
          return React.cloneElement(selectItem, {
            className: classNames(
              itemStyles.selectItem,
              selected === index && itemStyles.selected
            ),
            key: index,
            onClick: () => {
              onClose();
              onSetSelectedIndex(index);
            },
            onKeyDown: (event: React.KeyboardEvent<HTMLElement>) => {
              if (event.code === "Space") {
                onClose();
                onSetSelectedIndex(index);
                triggerRef.current?.focus();
              }
              if (event.code === "ArrowUp") {
                focusPreviousItem();
              }
              if (event.code === "ArrowDown") {
                focusNextItem();
              }
            },
            role: "listitem",
            tabIndex:
              index === focusedSelectItemIndex ? 0 : undefined,
          });
        } else {
          return null;
        }
      })}
    </div>
  );
};

SelectContainer.displayName = "SelectContainer";
