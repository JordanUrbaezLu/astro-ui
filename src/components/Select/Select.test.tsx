import { render, screen } from "@testing-library/react";
import { Select, SelectVariant, SelectSize } from "./Select";
import { SelectItem } from "./SelectItem";
import { axe } from "jest-axe";

describe("Select", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Select>
        <SelectItem>30</SelectItem>
      </Select>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Select>
        <SelectItem>30</SelectItem>
      </Select>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test.each<SelectVariant>(["primary", "secondary"])(
    "Should render variants correctly.",
    (variant) => {
      render(
        <Select label="select" variant={variant}>
          <SelectItem>30</SelectItem>
        </Select>
      );
      expect(screen.getByText("select")).toHaveClass(`${variant}Label`);
    }
  );

  test.each<SelectSize>(["large", "medium", "small"])(
    "Should render size correctly.",
    (size) => {
      render(
        <Select label="select" size={size}>
          <SelectItem>30</SelectItem>
        </Select>
      );
      expect(screen.getByText("select")).toHaveClass(`${size}Label`);
    }
  );

  test("Should render label correctly.", () => {
    render(
      <Select label="Select">
        <SelectItem>30</SelectItem>
      </Select>
    );
    expect(screen.getByText("Select")).toBeInTheDocument();
  });
});
