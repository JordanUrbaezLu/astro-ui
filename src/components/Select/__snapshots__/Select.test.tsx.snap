// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Should render component correctly. 1`] = `
<div
  class="selectContainer"
>
  <div
    class="container"
  >
    <div
      class="textFieldContainer"
    >
      <input
        aria-label="textfield"
        class="textField large inputContainer"
        id="void_textfield"
        required=""
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
  <svg
    class="icon"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M112 184l144 144 144-144"
      fill="none"
      stroke-linecap="square"
      stroke-miterlimit="10"
      stroke-width="48"
    />
  </svg>
</div>
`;
